File: argon-plugin/shared/src/main/scala/dev/argon/plugin/Backend.scala
package dev.argon.plugin

import dev.argon.compiler.*
import dev.argon.compiler.module.*
import dev.argon.compiler.tube.ArTubeC
import dev.argon.io.*
import dev.argon.options.*
import dev.argon.util.*
import zio.*

trait Backend[Options[_, _], Output[_, _], -R0, +E0] {
  def emitTube
  (context: Context { type Env <: R0; type Error >: E0 })
  (options: Options[context.Env, context.Error])
  (tube: ArTubeC with HasContext[context.type])
  : context.Comp[Output[context.Env, context.Error]]
}

File: plugins/java/api/src/dev/argon/plugin/api/Backend.java
package dev.argon.plugin.api;

import java.util.Map;
import java.util.Set;
import dev.argon.plugin.api.resource.Resource;
import dev.argon.plugin.api.tube.SerializedTube;

public interface Backend<E extends Exception, TOptions, TOutput> {
    TOutput emitModule(TOptions options, Set<Platform<E, ?, ?, ?>> platforms, SerializedTube<E> tube);
}
